pkg/infra/kubernetes/exec_unit.go

```diff
@@ -443,3 +428,62 @@ func (unit *HelmExecUnit) addEnvVarToPod(envVars core.EnvironmentVariables) ([]H
 
 	return values, nil
 }
+
+// upsertOnlyContainer ensures that there is exactly one container in the given slice, and that it's correctly
+// configured. Along the way, it will also generate the image placeholder value, which it then returns.
+//
+// If the provided containers slice is empty, this method will create a new container and inert it into the slice; this
+// modifies the call site's slice (which is why we pass in a pointer to a slice). Otherwise, we'll use the slice's
+// existing container, or return an error if there is more than one.
+//
+// To configure the container, we:
+//  1. set its image to a template for the generated placeholder value
+//  2. call configureContainer on it
+func (unit *HelmExecUnit) upsertOnlyContainer(containers *[]corev1.Container, cfg config.ExecutionUnit) (string, error) {
+	if len(*containers) > 1 {
+		return "", errors.New("too many containers in pod spec, don't know which to replace")
+	}
+	if len(*containers) == 0 {
+		*containers = append(*containers, corev1.Container{
+			Name: unit.Name,
+		})
+	}
+	container := &(*containers)[0]
+
+	value := GenerateImagePlaceholder(unit.Name)
+	container.Image = fmt.Sprintf("{{ .Values.%s }}", value)
+
+	if err := unit.configureContainer(container, cfg); err != nil {
+		return "", err
+	}
+
+	return value, nil
+}
+
+func (unit *HelmExecUnit) configureContainer(container *corev1.Container, cfg config.ExecutionUnit) error {
+	k8sCfg := cfg.GetExecutionUnitParamsAsKubernetes()
+
+	limits := make(map[corev1.ResourceName]any)
+	if k8sCfg.Limits.Cpu != nil {
+		limits[corev1.ResourceCPU] = k8sCfg.Limits.Cpu
+	}
+	if k8sCfg.Limits.Memory != nil {
+		limits[corev1.ResourceMemory] = k8sCfg.Limits.Memory
```

Comment:
in order for limits to actually do anything, we need to set the requests, otherwise it will never scale. We dont surface requests in config because we assume limits are 100% of requests for now so we can try to fully utilize nodes